Indicators/AdvanceDeclineRatio.cs

```diff
@@ -0,0 +1,145 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using QuantConnect.Data.Market;
+using QuantConnect.Logging;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// The advance-decline ratio (ADR) compares the number of stocks 
+    /// that closed higher against the number of stocks 
+    /// that closed lower than their previous day's closing prices.
+    /// </summary>
+    public class AdvanceDeclineRatio : TradeBarIndicator, IIndicatorWarmUpPeriodProvider
+    {
+        private IDictionary<string, TradeBar> _previousPeriod = new Dictionary<string, TradeBar>(StringComparer.OrdinalIgnoreCase);
+        private IDictionary<string, TradeBar> _currentPeriod = new Dictionary<string, TradeBar>(StringComparer.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AdvanceDeclineRatio"/> class
+        /// </summary>
+        public AdvanceDeclineRatio(string name) : base(name)
+        {
+
+        }
+
+        /// <summary>
+        /// Add tracking stock issue
+        /// </summary>
+        /// <param name="symbol">tracking stock issue</param>
+        public void AddStock(Symbol symbol)
+        {
+            if (!_currentPeriod.ContainsKey(symbol.Value))
+            {
+                _currentPeriod.Add(symbol.Value, null);
+            }
+        }
+
+        /// <summary>
+        /// Remove tracking stock issue
+        /// </summary>
+        /// <param name="symbol">tracking stock issue</param>
+        public void RemoveStock(Symbol symbol)
+        {
+            if (!_currentPeriod.ContainsKey(symbol.Value))
+            {
+                _currentPeriod.Remove(symbol.Value);
+            }
+        }
+
+        /// <summary>
+        /// Gets a flag indicating when this indicator is ready and fully initialized
+        /// </summary>
+        public override bool IsReady => Samples > 0;
+
+        /// <summary>
+        /// Required period, in data points, for the indicator to be ready and fully initialized.
+        /// </summary>
+        public int WarmUpPeriod { get; }
+
+        /// <summary>
+        /// Computes the next value of this indicator from the given state
+        /// </summary>
+        /// <param name="input">The input given to the indicator</param>
+        /// <returns>A new value for this indicator</returns>
+        protected override decimal ComputeNextValue(TradeBar input)
+        {
+            if (_currentPeriod.Any(t => t.Value != null) && (input.Time != _currentPeriod.First(t => t.Value != null).Value.Time))
```

Comment:
What if the indicator is feed with Minute resolution data, should it be checking for date change instead? Better the period should be an argument of the indicator and we use it to determine what previous and current are?

Reply:
we rely on consolidated data here.